ObjectDisposedException in FileSystemWatcher tests

Recently seeing occasional failures in the FileSystemWatcher tests with this exception:

http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_debug/834/consoleFull

05:02:06   Unhandled Exception: System.ObjectDisposedException: Safe handle has been closed
05:02:06      at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
05:02:06      at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
05:02:06      at Microsoft.Win32.Win32Native.SetEvent(SafeWaitHandle handle)
05:02:06      at System.Threading.EventWaitHandle.Set()
05:02:06      at Utility.<>c__DisplayClass8.<WatchForEvents>b__0(Object o, FileSystemEventArgs e)
05:02:06      at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(WatcherChangeTypes changeType, String name)
05:02:06      at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(Int32 action, String name)
05:02:06      at System.IO.FileSystemWatcher.CompletionStatusChanged(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer)
05:02:06      at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
05:02:06      at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)
05:02:07 d:\j\workspace\dotnet_corefx_windows_debug\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00063\lib\tests.targets(125,5): warning MSB3073: The command "corerun.exe xunit.console.netcore.exe System.IO.FileSystem.Watcher.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -532462766. [d:\j\workspace\dotnet_corefx_windows_debug\src\System.IO.FileSystem.Watcher\tests\System.IO.FileSystem.Watcher.Tests.csproj]
05:02:07 d:\j\workspace\dotnet_corefx_windows_debug\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00063\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.IO.FileSystem.Watcher.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_debug\src\System.IO.FileSystem.Watcher\tests\System.IO.FileSystem.Watcher.Tests.csproj]
05:02:07   System.IO.FileSystem.DriveInfo.Tests -> d:\j\workspace\dotnet_corefx_windows_debug\bin\Windows_NT.AnyCPU.Debug\System.IO.FileSystem.DriveInfo.Tests\System.IO.FileSystem.DriveInfo.Tests.dll

/cc @chcosta @sokket

--- a/src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Deleted.cs
+++ b/src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Deleted.cs
@@ -86,9 +86,10 @@ public partial class FileSystemWatcher_4000_Tests
     {
         using (var dir = Utility.CreateTestDirectory())
         using (var watcher = new FileSystemWatcher())
-        using (AutoResetEvent createOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Created))
-        using (AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Deleted))
         {
+            AutoResetEvent createOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Created); // not "using" to avoid race conditions with FSW callbacks
+            AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Deleted);
+
             watcher.Path = Path.GetFullPath(dir.Path);
             watcher.Filter = "*";
             watcher.IncludeSubdirectories = true;
@@ -125,9 +126,10 @@ public partial class FileSystemWatcher_4000_Tests
     {
         using (var dir = Utility.CreateTestDirectory())
         using (var watcher = new FileSystemWatcher())
-        using (AutoResetEvent createOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Created))
-        using (AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Deleted))
         {
+            AutoResetEvent createOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Created); // not "using" to avoid race conditions with FSW callbacks
+            AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Deleted);
+
             watcher.Path = Path.GetFullPath(dir.Path);
             watcher.Filter = "*";
             watcher.IncludeSubdirectories = true;

--- a/src/System.IO.FileSystem.Watcher/tests/Utility/Utility.cs
+++ b/src/System.IO.FileSystem.Watcher/tests/Utility/Utility.cs
@@ -102,15 +102,16 @@ public static class Utility
     }
 
     public static void TestNestedDirectoriesHelper(
-    WatcherChangeTypes change,
-    Action<AutoResetEvent, TemporaryTestDirectory> action,
-    NotifyFilters changeFilers = NotifyFilters.LastWrite | NotifyFilters.FileName | NotifyFilters.DirectoryName)
+        WatcherChangeTypes change,
+        Action<AutoResetEvent, TemporaryTestDirectory> action,
+        NotifyFilters changeFilers = NotifyFilters.LastWrite | NotifyFilters.FileName | NotifyFilters.DirectoryName)
     {
         using (var dir = Utility.CreateTestDirectory())
         using (var watcher = new FileSystemWatcher())
-        using (AutoResetEvent createdOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Created))
-        using (AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, change))
         {
+            AutoResetEvent createdOccured = Utility.WatchForEvents(watcher, WatcherChangeTypes.Created); // not "using" to avoid race conditions with FSW callbacks
+            AutoResetEvent eventOccured = Utility.WatchForEvents(watcher, change);
+
             watcher.Path = Path.GetFullPath(dir.Path);
             watcher.Filter = "*";
             watcher.NotifyFilter = changeFilers;

